File: java-multi-thread-test/src/main/java/com/tistory/devs0n/ProductRepository.java
package com.tistory.devs0n;

import java.util.HashMap;
import java.util.Map;

public class ProductRepository {
    private Map<String, Product> products = new HashMap<>();

    public Product save(final Product product) {
        final var productClone = product.clone();
        this.products.put(product.id, productClone);
        return product;
    }

    public Product getById(final String productId) {
        final var product = this.products.get(productId);

        if (product == null) {
            throw new IllegalArgumentException("Product not found");
        }

        return product.clone();
    }
}

File: spring-redisson-lock/src/main/kotlin/com/tistory/devs0n/lock/product/ProductRepository.kt
package com.tistory.devs0n.lock.product

interface ProductRepository {
    fun save(product: Product): Product

    fun saveAll(products: List<Product>): List<Product>

    fun findByNumber(number: Long): Product?

    fun findAllByNumberIn(numbers: List<Long>): List<Product>
}
